packages/medusa/src/services/draft-order.js

```diff
@@ -0,0 +1,433 @@
+      const createdCart = await this.cartService_
+        .withTransaction(manager)
+        .create({ type: "draft_order", ...rest })
+
+      const draftOrder = draftOrderRepo.create({ cart_id: createdCart.id })
+      const result = await draftOrderRepo.save(draftOrder)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(DraftOrderService.Events.CREATED, {
+          id: result.id,
+        })
+
+      let shippingMethods = []
+      let profiles = []
+      if (shippingRequired) {
+        for (const method of shipping_methods) {
+          const m = await this.shippingOptionService_
+            .withTransaction(manager)
+            .createShippingMethod(method.option_id, method.data, {
+              cart: createdCart,
+            })
+
+          shippingMethods.push(m)
+        }
+
+        profiles = shippingMethods.map(
+          ({ shipping_option }) => shipping_option.profile_id
+        )
+      }
+
+      for (const item of items) {
+        if (item.variant_id) {
+          const line = await this.lineItemService_
+            .withTransaction(manager)
+            .generate(
+              item.variant_id,
+              data.region_id,
+              item.quantity,
+              item.metadata
+            )
+
+          const variant = await this.productVariantService_
+            .withTransaction(manager)
+            .retrieve(item.variant_id)
+          const itemProfile = variant.product.profile_id
+
+          let hasShipping = true
+
+          // if shipping is required, ensure items can be shipped
+          if (shippingRequired) {
+            hasShipping = profiles.includes(itemProfile)
+          }
+
+          await this.lineItemService_.withTransaction(manager).create({
+            cart_id: createdCart.id,
+            has_shipping: hasShipping,
+            ...line,
+          })
+        } else {
+          // custom line items can be added to a draft order
+          await this.lineItemService_.withTransaction(manager).create({
+            cart_id: createdCart.id,
+            has_shipping: true,
+            title: item.title || "Custom item",
+            allow_discounts: false,
+            unit_price: item.unit_price || 0,
+            quantity: item.quantity,
+          })
+        }
+      }
+
+      return result
+    })
+  }
+
+  /**
+   * Registers a system payment and creates an order
+   * @param {string} doId - id of draft order to register payment for
+   * @return {Promise<Order>} the created order
+   */
+  async registerSystemPayment(doId) {
+    return this.atomicPhase_(async manager => {
+      const draftOrder = await this.retrieve(doId)
+
+      const draftOrderCart = await this.cartService_
+        .withTransaction(manager)
+        .retrieve(draftOrder.cart_id, {
+          select: ["total"],
+          relations: ["discounts", "shipping_methods", "region", "items"],
+        })
+
+      const orderRepo = manager.getCustomRepository(this.orderRepository_)
```

Comment:
We should be calling some `OrderService` methods instead of using the order repository directly

Reply:
Potentially a `OrderService.createFromDraft`

Reply:
Another reason for this is that if we create it directly the order won't emit `order.placed` event which is necessary for many plugins to work.

Reply:
I agree.. This is a good idea. Hope about reusing `orderService.createFromCart`? We don't really have to differentiate a draft order cart from a default cart, when it comes to order creation.

We could potentially do all the work in the route - something along the lines of:
```
await entityManager.transaction(async manager => {
      const draftOrder = await draftOrderService
        .withTransaction(manager)
        .retrieve(id)

      const cart = await cartService
        .withTransaction(manager)
        .retrieve(draftOrder.cart_id, {
          select: ["total"],
          relations: ["discounts", "shipping_methods", "region", "items"],
        })

      await paymentProviderService
        .withTransaction(manager)
        .createSession("system", cart)

      await cartService
        .withTransaction(manager)
        .authorizePayment(draftOrder.cart_id)

      const result = await orderService
        .withTransaction(manager)
        .createFromCart(draftOrder.cart_id)

      const order = await orderService.retrieve(result.id, {
        relations: defaultOrderRelations,
        select: defaultOrderFields,
      })

      res.status(200).json({ order })
})
```

Reply:
Perhaps introduce `idempotency phases` as well